k2/csrc/array.h

```diff
@@ -1,69 +1,113 @@
 // k2/csrc/array.h
 
 // Copyright (c)  2020  Xiaomi Corporation (author: Daniel Povey)
+//                      Fangjun Kuang (csukuangfj@gmail.com)
 
 // See ../../LICENSE for clarification regarding multiple authors
 
 #ifndef K2_CSRC_ARRAY_H_
 #define K2_CSRC_ARRAY_H_
 
-#include <functional>
-#include <limits>
-#include <memory>
+#include <type_traits>
 #include <vector>
 
+#include "glog/logging.h"
+
 namespace k2 {
 
 /*
-   We will use e.g. StridedPtr<int32_t, T> when the stride is not 1, and
+   We will use e.g. StridedPtr<T*, int32_t> when the stride is not 1, and
    otherwise just T* (which presumably be faster).
 */
-template <typename I, typename T>
+template <typename Ptr, typename I = int32_t>
```

Comment:
I think we should pass `T` here instead of `Ptr/T*` as the template parameter, generally, it's ok if user want to pass `T* *` or something like `StridePtr` as the template paramter (even though we may not need those things).

Reply:
@qindazhu If we pass the value type we can't use StridedPtr which is actually needed.
I think @csukuangfj  has it right.

Reply:
Oh I understand now.  I think both ways would be acceptable but the way @csukuangfj has done it may be more general.  Just as long as it's clearly documented, the way it's done here is OK with me.